doc/source/user_guide/data_types.rst

```diff
@@ -176,7 +181,7 @@ a custom function that requires a particular dtype, you should call one of the
 dtype conversion functions (here, ``func1`` and ``func2`` are ``skimage``
 functions)::
 
-   >>> from skimage import img_as_float
+   >>> from skimage.util import img_as_float
```

Comment:
Why these changes?  I think those function remain exposed on the default namespace, without plans to change that any time soon or ever.

Reply:
@stefanv I object to the second clause in that sentence. =P 

I want this to be deprecated in the same way that `np.matrix` was "deprecated" for a long time before any messages appeared. ie, discouraged. For me it makes much more sense for these functions to be in `util` than "polluting" the top-level domain, which is otherwise only modules and the `data_dir` variable.

Reply:
I should have been clearer, I guess: without your plans having been made known to the rest of us? :)

Seriously, though, those functions are strewn throughout *all* our lecture materials, so it would be quite painful to get rid of it.  Now, w.r.t. discouragement, I could go for that, but do you have a specific rationale for why `skimage.util.img_as_float` is better than `skimage.img_as_float`, other than the slight aesthetic benefit of having only submodules in the namespace?  I guess then we'll also need to get rid of `skimage.test()`, which leaves only submodules + `__version__`.

Reply:
I once listened to a talk that was advocate against this kind of nested API. They were saying that we should strive for a "flat" API design. not sure how applicable that is to skimage.

a flat API can be easier to learn, at least easier to remember what to import from.

Reply:
I was under the impression that this idea had been lightly discussed elsewhere, but I can't find any evidence of that, so, sorry about that.

I don't consider "this is the way we've always done it" to be a big deal, especially in the context of a soft deprecation. There is zero urgency in this removal, it is just something that I think we should do "right" going forward, and especially in our own code base. There's outdated tutorials for everything online, and this should not hold us back, any more than "there's lots of tutorials and books written in Python 2" should hold back the transition to 3. Now, I was auditing the use of these functions in our codebase so I took the opportunity to update these. (And note that we had both conventions floating about in our codebase, which is a jarring inconsistency.)

I actually think `test` and `lookfor` and maybe even `data_dir` should stay in the top level as being "package-wide" things. Is `img_as_float` something that refers to our package as a whole, or a utility function? For me it's a slam dunk for the latter. Put another way: does `img_as_float` belong in the same bucket as `lookfor` and `test`?

Reply:
Perhaps what bothers me is the nondescript `util` name, which will not help users discover these.  They could potentially live in `io`, or in their own submodule.

You are arguing that these functions are not-package wide, which I agree with.  But another perspective could be that they belong there because they are invoked on every single pipeline, and should therefore be highly visible.

Hopefully they will be needed less as we embrace float images without defined ranges, wherever possible.

Reply:
Hey I'm the one who argued for a flat namespace to help discoverability and was vetoed. =P

More seriously, as you say, I hope that long term they are not so indispensable.

Finally, perhaps a new module, `skimage.dtypes`? Or somesuch?

Reply:
It seems that maybe you should restrict the scope of the PR. I don't think there is much resistance to any of the proposed changes, but it seems that together, they are hard to justify at once as some are cosmetic, and some are behavioural.